zerver/lib/actions.py

```diff
@@ -833,6 +833,30 @@ def active_humans_in_realm(realm: Realm) -> Sequence[UserProfile]:
     return UserProfile.objects.filter(realm=realm, is_active=True, is_bot=False)
 
 
+def update_users_in_members_and_full_members_groups(realm: Realm) -> None:
+    full_members_system_group = UserGroup.objects.get(
+        realm=realm, name="@role:fullmembers", is_system_group=True
+    )
+    members_system_group = UserGroup.objects.get(
+        realm=realm, name="@role:members", is_system_group=True
+    )
+
+    full_member_group_users = set(full_members_system_group.direct_members.all())
+    new_members = [user for user in full_member_group_users if user.is_provisional_member]
+
+    member_group_users = set(members_system_group.direct_members.all())
+    members_excluding_full_members = list(member_group_users - full_member_group_users)
+    new_full_members = [
+        user for user in members_excluding_full_members if not user.is_provisional_member
+    ]
```

Comment:
Does this logic work consistently?  I have memories, possibly very outdated, of Django objects not behaving how one would except when placed in sets in some situations, and we generally try to work with IDs in instead for that reason.

Reply:
The test works fine. I am not sure how to check whether it works consistently or not and am not aware of any case where it could fail. Is the problem related to sets only and we can use lists instead. We would surely need objects to check `is_provisional_member` and thus ID is not enough.

Reply:
Maybe we can use `ids` for calculating `members_excluding_full_members`.

Reply:
If we want to completely avoid use of list/set of objects, we can directly use `date_joined` value in query. Code will look something like -
```
full_member_group_user_ids = full_members_system_group.direct_members.all().values_list("id", flat=True)

min_age = datetime.timedelta(days=realm.waiting_period_threshold)
new_members = full_members_system_group.direct_members.filter(
    date_joined__gt=timezone_now() - min_age
)
new_full_members = members_system_group.direct_members.filter(
    date_joined__lt=timezone_now() - min_age
).exclude(id__in=full_member_group_user_ids)

if len(new_members) > 0:
   remove_members_from_user_group(full_members_system_group, new_members)

if len(new_full_members) > 0:
   bulk_add_members_to_user_group(full_members_system_group, new_full_members)
```

Reply:
OK, I've confirmed that the `set` approach will work correctly if we're careful about ensuring the sets we work with have a consistent Django model type with an appropriate primary key.

See e.g. https://github.com/django/django/commit/4668c142dce77c6f29fb75532c1acfa1b2d322ff for background.

Reply:
(I'll drop the bonus commit)